notes/advanced-data-types.md

```diff
@@ -0,0 +1,530 @@
+
+```
+/*
+enum enumName {
+    constant1 = value1,
+    constant2 = value2,
+    ...
+    constantN = valueN
+};
+*/
+
+enum weekdays {
+    Sunday = 0,
+    Monday = 1,
+    Tuesday = 2,
+    Wednesday = 3,
+    Thursday = 4,
+    Friday = 5,
+    Saturday = 6
+};
+```
+<div style="text-align: center"> Figure 4b - Enum Declaration </div>
+
+Now, if you want specific values for each constant, you can use this `= integer` notation, as shown below:
+```
+enum painMeter {
+    noPain = 0,
+    somePain = 3,
+    pain = 5,
+    extremePain = 7,
+    CSE351 = 10
+};
+```
+<div style="text-align: center"> Figure 4c - Enum Declaration </div>
+
+You can also do a mix of both
+
+```
+enum HuskySatTwo {
+    LOST,
+    ReactionWheels = 4,
+    Magnetorquers
+    FOUND,
+    Propulsion = 10
+};
+```
+<div style="text-align: center"> Figure 4d - Enum Declaration </div>
+
+- `LOST` has no intitialized value, so its value is 0
+- `ReactionWheels` has an initialized value, so its value is 4
+- `Magnetorquers` has no initialized value, so it is one more than the last, or 5
+- `FOUND` also has no initialized value, so it is 6
+- `Propulsion` does, so it is 10
+
+After an enum is declared, you can then make variables out of it using the following notation, which is similar to the declaration of a struct
+
+```
+// enumName enumVarName = constant;
+
+weekday day = Wednesday;
+painMeter pain;
+pain = extremePain;
+```
+<div style="text-align: center"> Figure 4e - Enum Initialization </div>
+
+## Enumeration Classes (`enum class`)
+
+However, once declared, enums will not allow for any redefinition of any of its constants while you are in the enum's scope. An enum's scope ranges from its declaration to the `}` brackets that contain it (Or the end of the file), and this applies to its fields as well. The below demonstrates a few errors about this:
+
+```
+enum People = {Good, Bad}; // Begin People Scope
+
+int main() {
+    enum Others = {Good, Meh}; // Begin Others Scope
+    // Error: Good has already been defined by People
+    
+    int Meh = 0;
+    // Error: Bad is already a constant
+} // End Others Scope
+
+enum Professors = {Okay, Meh} // Begin Professors Scope
+// Okay: the enum Others is out of scope of Professors
+
+// End People Scope
+// End Professors Scope
+```
+<div style="text-align: center"> Figure 5a - Enum Errors </div>
+
+Another bad thing is attempting to compare different enums when you don't want it. This is because C++ is not type-safe, so it automatically convert the `enum` constants to integers in a process known as *implicit conversion*. The following program functions, but doesn't conceptually make any sense:
+
+```
+weekday day = Friday;
+painMeter painRating = pain;
+if(painRating == painRating) {
```

Comment:
Wait you don't seem to be using `day`, just comparing `painRating` with itself

Reply:
:(